File: puremacro/vfi/egm.py
"""Endogenous Grid Method (Carroll 2006) for the one-asset income-fluctuation problem.

Solves  max E sum beta^t u(c),  u(c) = c^(1-gamma)/(1-gamma),  subject to
    c + a' = (1+r) a + y(z),    a' >= a_grid[0]   (borrowing constraint),
with z an AR(1) Markov shock (P_z) and income y(z). EGM inverts the Euler
equation on a grid of NEXT-period assets instead of maximising over a' on a
discrete grid -- O(n) per iteration, and the policy is continuous (interpolated)
rather than pinned to grid nodes. Returns continuous c(a,z) and a'(a,z) policies.
numpy-only (the EGM interpolation is not a GPU/backend path).
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np


@dataclass(frozen=True)
class EGMSolution:
    """Continuous consumption and next-asset policies from EGM (numpy)."""
    c: np.ndarray            # (n_a, n_z) consumption policy
    aprime: np.ndarray       # (n_a, n_z) next-asset policy (values, >= a_grid[0])
    n_iter: int
    sup_norm: float


def solve_egm(a_grid, z_grid, income, P_z, *, beta, r, gamma,
              tol: float = 1e-9, max_iter: int = 10_000):
    """EGM solve of the CRRA income-fluctuation problem; see module docstring.

    ``income`` is y(z), shape (n_z,). Returns an EGMSolution with continuous
    policies on ``a_grid``. (The top of ``a_grid`` should be high enough that the
    saving policy is interior there; np.interp clamps above the endogenous range.)
    """
    a = np.asarray(a_grid, dtype=float)
    inc = np.asarray(income, dtype=float)
    P = np.asarray(P_z, dtype=float)
    n_a, n_z = a.size, np.asarray(z_grid).shape[0]
    if not (0.0 < beta < 1.0):
        raise ValueError(f"beta must be in (0,1); got {beta}")
    if r <= -1.0:
        raise ValueError(f"r must be > -1; got {r}")
    if gamma <= 0.0:
        raise ValueError(f"gamma must be > 0; got {gamma}")
    if inc.shape != (n_z,):
        raise ValueError(f"income must have shape ({n_z},); got {inc.shape}")
    if not np.all(np.diff(a) > 0):
        raise ValueError("a_grid must be strictly increasing")
    if P.shape != (n_z, n_z) or not np.allclose(P.sum(axis=1), 1.0, atol=1e-8) or np.any(P < -1e-12):
        raise ValueError("P_z must be (n_z,n_z) and row-stochastic")
    R = 1.0 + r
    a_min = a[0]
    coh = R * a[:, None] + inc[None, :]                  # cash-on-hand (n_a, n_z)
    c = np.maximum(coh - a_min, 1e-10)                   # initial guess: save the minimum
    sup = np.inf
    for it in range(1, max_iter + 1):
        Emu = (c ** (-gamma)) @ P.T                      # E_{z'|z}[u'(c')], indexed (a', z)
        rhs = beta * R * Emu
        c_endog = rhs ** (-1.0 / gamma)                  # current c if next assets = a-node
        a_endog = (c_endog + a[:, None] - inc[None, :]) / R   # endogenous current assets
        c_new = np.empty((n_a, n_z))
        for zi in range(n_z):
            ae = a_endog[:, zi]
            ce = c_endog[:, zi]
            c_un = np.interp(a, ae, ce)                  # unconstrained (clamped outside)
            c_con = coh[:, zi] - a_min                   # constrained: a' = a_min
            c_new[:, zi] = np.where(a <= ae[0], c_con, c_un)
        c_new = np.maximum(c_new, 1e-10)
        sup = float(np.max(np.abs(c_new - c)))
        c = c_new
        if sup < tol:
            break
    else:
        raise RuntimeError(
            f"solve_egm did not converge in {max_iter} iterations "
            f"(sup-norm {sup:.3e} > tol {tol:.1e})"
        )
    aprime = np.clip(coh - c, a_min, a[-1])
    c = coh - aprime   # re-impose the budget after clamping a' to the grid range
    return EGMSolution(c=c, aprime=aprime, n_iter=it, sup_norm=sup)


__all__ = ["solve_egm", "EGMSolution"]

File: matlab/+puremacro/+vfi/egm.m
function res = egm(a_grid, z_grid, P, beta, r_rate, gamma_risk, max_iter, tol)
% EGM Endogenous Grid Method (Carroll 2006) for ultra-fast $O(N_a N_z)$ VFI.
%
%   RES = puremacro.vfi.egm(A_GRID, Z_GRID, P, BETA, R_RATE, GAMMA_RISK, MAX_ITER, TOL)
%
%   Solves: u'(c_t) = beta * (1 + r) * E_t [ u'(c_{t+1}) ]
%   Without grid search over a_prime!

    if nargin < 5 || isempty(r_rate)
        r_rate = 0.04;
    end
    if nargin < 6 || isempty(gamma_risk)
        gamma_risk = 2.0;
    end
    if nargin < 7 || isempty(max_iter)
        max_iter = 1000;
    end
    if nargin < 8 || isempty(tol)
        tol = 1e-7;
    end

    a_grid = a_grid(:);
    z_grid = z_grid(:);
    Na = length(a_grid);
    Nz = length(z_grid);

    R_fac = 1.0 + r_rate;

    % Initial policy guess: c(a, z) = r * a + z
    c_policy = zeros(Na, Nz);
    for k = 1:Nz
        c_policy(:, k) = max(r_rate * a_grid + z_grid(k), 1e-4);
    end

    t_start = tic;
    iter = 0;
    dist = 1.0;

    while dist > tol && iter < max_iter
        iter = iter + 1;
        c_old = c_policy;

        % Marginal utility u'(c) = c^(-gamma)
        mu_next = c_old .^ (-gamma_risk);

        % Expected marginal utility E_t [ u'(c_{t+1}) ] = mu_next * P'
        emu = mu_next * P';

        % Endogenous consumption choice c_endog = (beta * R * emu)^(-1 / gamma)
        c_endog = (beta * R_fac * emu) .^ (-1.0 / gamma_risk);

        % Endogenous asset grid: a_endog = (c_endog + a_grid' - z_grid') / R
        % Re-interpolate onto exogenous a_grid
        c_policy_new = zeros(Na, Nz);
        for k = 1:Nz
            a_endog = (c_endog(:, k) + a_grid - z_grid(k)) / R_fac;

            % Monotone interpolation on exogenous a_grid
            for i = 1:Na
                a_target = a_grid(i);
                if a_target <= a_endog(1)
                    % Borrowing constraint binds
                    c_policy_new(i, k) = R_fac * a_target + z_grid(k) - a_grid(1);
                else
                    c_policy_new(i, k) = interp1(a_endog, c_endog(:, k), a_target, 'linear', 'extrap');
                end
            end
        end

        c_policy_new = max(c_policy_new, 1e-6);
        dist = max(abs(c_policy_new(:) - c_old(:)));
        c_policy = c_policy_new;
    end

    elapsed = toc(t_start);

    % Implied asset policy a'(a, z) = R * a + z - c(a, z)
    policy_a = zeros(Na, Nz);
    for k = 1:Nz
        policy_a(:, k) = R_fac * a_grid + z_grid(k) - c_policy(:, k);
        policy_a(:, k) = max(policy_a(:, k), a_grid(1));
    end

    res.c_policy = c_policy;
    res.policy_a = policy_a;
    res.n_iter = iter;
    res.elapsed = elapsed;
end
